wp-lazy-loading.php

```diff
@@ -118,57 +120,153 @@ function wp_lazy_loading_enabled( $tag_name, $context ) {
+
+	foreach ( $images as $image => $attachment_id ) {
+		$filtered_image = $image;
+
+		// Add 'srcset' and 'sizes' attributes if applicable.
+		if ( $attachment_id > 0 && false === strpos( $filtered_image, ' srcset=' ) ) {
+			$filtered_image = wp_image_tag_add_srcset_and_sizes_attr( $filtered_image, $context, $attachment_id );
+		}
+
+		// Add 'loading' attribute if applicable.
+		if ( $add_loading_attr && false === strpos( $filtered_image, ' loading=' ) ) {
+			$filtered_image = wp_image_tag_add_loading_attr( $filtered_image, $context );
+		}
+
+		if ( $filtered_image !== $image ) {
+			$content = str_replace( $image, $filtered_image, $content );
+		}
+	}
+
+	return $content;
+}
+
+/**
+ * Adds `loading` attribute to an existing `img` HTML tag.
+ *
+ * @since (TBD)
+ *
+ * @param string $image   The HTML `img` tag where the attribute should be added.
+ * @param string $context Additional context to pass to the filters.
+ * @return string Converted `img` tag with `loading` attribute added.
+ */
+function wp_image_tag_add_loading_attr( $image, $context ) {
+	/**
+	 * Filters the `loading` attribute value. Default `lazy`.
+	 *
+	 * Returning `false` or an empty string will not add the attribute.
+	 * Returning `true` will add the default value.
+	 *
+	 * @since (TBD)
+	 *
+	 * @param string $value   The 'loading' attribute value, defaults to `lazy`.
+	 * @param string $image   The HTML 'img' element to be filtered.
+	 * @param string $context Additional context about how the function was called or where the img tag is.
+	 */
+	$value = apply_filters( 'wp_image_tag_add_loading_attr', 'lazy', $image, $context );
+
+	if ( $value ) {
+		if ( ! in_array( $value, array( 'lazy', 'eager' ), true ) ) {
+			$value = 'lazy';
+		}
+
+		return str_replace( '<img', '<img loading="' . $value . '"', $image );
+	}
+
+	return $image;
+}
+
+/**
+ * Adds `srcset` and `sizes` attributes to an existing `img` HTML tag.
+ *
+ * @since (TBD)
+ *
+ * @param string $image         The HTML `img` tag where the attribute should be added.
+ * @param string $context       Additional context to pass to the filters.
+ * @param int    $attachment_id Image attachment ID.
+ * @return string Converted 'img' element with 'loading' attribute added.
+ */
+function wp_image_tag_add_srcset_and_sizes_attr( $image, $context, $attachment_id ) {
+	/**
+	 * Filters whether to add the `srcset` and `sizes` HTML attributes to the img tag. Default `true`.
+	 *
+	 * Returning anything else than `true` will not add the attributes.
+	 *
+	 * @since (TBD)
+	 *
+	 * @param bool   $value         The filtered value, defaults to `true`.
+	 * @param string $image         The HTML `img` tag where the attribute should be added.
+	 * @param string $context       Additional context about how the function was called or where the img tag is.
+	 * @param int    $attachment_id The image attachment ID.
+	 */
+	$add = apply_filters( 'wp_image_tag_add_srcset_and_sizes_attr', true, $image, $context, $attachment_id );
```

Comment:
I'm not sure whether we should introduce a filter like this. I get the idea and am generally not opposed, but this is an unrelated change and would raise further discussion points that don't actually help what we want to achieve here.

I'd say let's simplify this function to only be the part within the below if clause.

Reply:
Yeah, still on the fence about it too.

Having a filter somewhat matches `wp_image_tag_add_loading_attr()` (and eventually `wp_iframe_tag_add_loading_attr()`). It also lets plugin disable srcset and sizes depending on `$context`, which might be useful in some edge cases but not strictly necessary.

Not having a filter makes introducing `wp_image_tag_add_srcset_and_sizes_attr()` somewhat questionable as the same code may as well be inside the loop where that function is called...

Perhps can keep it for now and revisit when making the core patch? Then there will likely be more people looking and helping :)

Reply:
Actually revisiting this, we should probably have this filter if we make the other change from above (supporting more than `the_content` here). With this filter, we'll allow developers to opt out for other contexts than `the_content` (not sure anybody will wanna do that, but at least it allows someone to get back to the previous behavior if they have some random reason to do so).